### PR TITLE
[9.x] Fix multiple calls to class cast mutators

### PR DESCRIPTION
**Pull request by zupolgec**
Solves #41533. Class casts (`Attribute` included) allow the developer to use a more expressive syntax like `$model->address->lineOne = '117 Spencer St.'` avoiding assigning the object to a variable and then updating the attribute to trigger the mutator.

What the framework currently does to allow this to work is "if a mutated attribute has already been accessed once, every time I access any attribute (through `getAttributes`), call the mutator to sync the changes made to the object back to the model attributes".

My tentative solution to this issue is: every time an attribute or class mutated attribute is accessed, consider that it could potentially also lead to an object modification. So flag this attribute as "potentially altered" and call the mutator on first occasion, the remove the flag until next access.

**EDIT**

This performance check to avoid multiple calls doesn't allow anymore to modify the object by itself but you need to access the object through the model.

So given
```
$address = new AttributeCastAddress('110 Kingsbrook St.', 'My Childhood House');
$model->address = $address
```
you should do
```
$model->address->lineOne = '117 Spencer St.';
```
not
```
$address->lineOne = '117 Spencer St.';
```

I guess that this introduces a breaking change and then this PR should target the next major version.

### CODE REVIEW

**Comment by linaspasv** [comment]
Unfortunately this does not solve the main problem. The main cause of too many getAttributes calls is because of irrational and expensive multiple uses of getDirty method during the save process.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

File: tests/Integration/Database/DatabaseEloquentModelAttributeCastingTest.php [hunk 1]
```diff
@@ -49,8 +49,8 @@ public function testBasicCustomCasting()
 
         $model = new TestEloquentModelWithAttributeCast;
 
-        $model->address = $address = new AttributeCastAddress('110 Kingsbrook St.', 'My Childhood House');
-        $address->lineOne = '117 Spencer St.';
+        $model->address = new AttributeCastAddress('110 Kingsbrook St.', 'My Childhood House');
+        $model->address->lineOne = '117 Spencer St.';
```
Comment (taylorotwell) [inline tests/Integration/Database/DatabaseEloquentModelAttributeCastingTest.php:53]: Why did you change this test?

Comment (zupolgec) [reply tests/Integration/Database/DatabaseEloquentModelAttributeCastingTest.php:53]: Apologies, I was sure I had added an edit to the PR description highlighting that by introducing this performance check to avoid multiple calls you then can't modify the object by itself but you need to access the object through the model.

So given
```
$address = new AttributeCastAddress('110 Kingsbrook St.', 'My Childhood House');
$model->address = $address
```
you should do
```
$model->address->lineOne = '117 Spencer St.';
```
not
```
$address->lineOne = '117 Spencer St.';
```

I guess that this introduces a breaking change and then this PR should target the next major version.

**Comment by zupolgec** [comment]
Yep, you're right (as I clarified in the PR description now). It's a use case I didn't think of at first, since I normally access attributes through the model. Probably just flagging every attribute with `withoutObjectCaching()` is enough if you want to save a mutator call every time an attribute is accessed.

**Comment by zupolgec** [comment]
@taylorotwell could perhaps this still be valid for AttributeMutated attributes? No need to update those every time, since they can only be accessed through the model.

**Comment by taylorotwell** [comment]
No - I think it causes the same problem with those as well. I don't think there is a guaranteed way to solve this other than the current implementation to be honest.
